### PR TITLE
feat: update to R17 openapi 

### PR DESCRIPTION
**Pull request by donald1218**

### CODE REVIEW

File: internal/gmm/handler.go [hunk 1]
```diff
@@ -300,13 +298,13 @@ func forward5GSMMessageToSMF(
...
 ) error {
-	smContextUpdateData := models.SmContextUpdateData{
+	smContextUpdateData := models.SmfPduSessionSmContextUpdateData{
 		N1SmMsg: &models.RefToBinaryData{
 			ContentId: "N1SmMsg",
 		},
 	}
 	smContextUpdateData.Pei = ue.Pei
-	smContextUpdateData.Gpsi = ue.Gpsi
+	// smContextUpdateData.Gpsi = ue.Gpsi
```
Comment (ian60509) [inline internal/gmm/handler.go:307]: This line should be removed.

File: internal/ngap/handler_test.go [hunk 2]
```diff
@@ -79,7 +79,7 @@ func NewAmfContext(amfCtx *amf_context.AMFContext) {
 		SBIPort:      8000,
 		ServedGuamiList: []models.Guami{
 			{
-				PlmnId: &models.PlmnId{
+				PlmnId: &models.PlmnIdNid{
```
[the remark this answers is not in the document]
Comment (donald1218) [reply internal/ngap/handler_test.go:82]: In TS 29.571
Guami's type is PlmnIdNid not PlmnId 
image

**Review by ian60509 — APPROVED** [review]

**Review by andy89923 — COMMENTED** [review]
`Nnrf_NFDiscovery.SearchNFInstancesParamOpts`'s `TargetNfType` must be non nil.
This PR miss this at multiple places.

File: internal/sbi/consumer/udm_service.go [hunk 3]
```diff
@@ -105,37 +96,30 @@ func (s *nudmService) SDMGetAmData(ue *amf_context.AmfUe) (problemDetails *model
...
-		ue.Gpsi = data.Gpsis[0] // TODO: select GPSI
-	} else if httpResp != nil {
-		if httpResp.Status != localErr.Error() {
-			err = localErr
-			return problemDetails, err
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		problemDetails = &problem
+		ue.AccessAndMobilitySubscriptionData = &data.AccessAndMobilitySubscriptionData
+		ue.Gpsi = data.AccessAndMobilitySubscriptionData.Gpsis[0] // TODO: select GPSI
```
Comment (andy89923) [inline internal/sbi/consumer/udm_service.go:116]: Check the size before reading! (Prevent Panic)

File: internal/sbi/consumer/nrf_service.go [hunk 4]
```diff
@@ -247,54 +251,52 @@ func (s *nnrfService) SendRegisterNFInstance(nrfUri, nfInstanceId string, profil
...
+		case <-ctx.Done():
+			return "", "", fmt.Errorf("context done")
+		default:
+			res, err = client.NFInstanceIDDocumentApi.RegisterNFInstance(ctx, registerNFInstanceRequest)
+			if err != nil || res == nil {
+				// TODO : add log
+				logger.ConsumerLog.Errorf("AMF register to NRF Error[%s]", err.Error())
+				time.Sleep(2 * time.Second)
+				continue
 			}
```
Comment (pf-lin) [inline internal/sbi/consumer/nrf_service.go:272]:
```suggestion
			res, err = client.NFInstanceIDDocumentApi.RegisterNFInstance(ctx, registerNFInstanceRequest)
			if err != nil || res == nil {
				// TODO : add log
				logger.ConsumerLog.Errorf("AMF register to NRF Error[%s]", err.Error())
				time.Sleep(2 * time.Second)
				continue
			}
			nf = res.NrfNfManagementNfProfile
```

File: internal/context/amf_ue.go [hunk 5]
```diff
@@ -684,9 +683,9 @@ func (ue *AmfUe) CopyDataFromUeContextModel(ueContext models.UeContext) {
 		if ue.AccessAndMobilitySubscriptionData == nil {
 			ue.AccessAndMobilitySubscriptionData = new(models.AccessAndMobilitySubscriptionData)
 		}
-		if ue.AccessAndMobilitySubscriptionData.SubscribedUeAmbr == nil {
-			ue.AccessAndMobilitySubscriptionData.SubscribedUeAmbr = new(models.AmbrRm)
-		}
+		// if ue.AccessAndMobilitySubscriptionData.SubscribedUeAmbr == nil {
+		// 	ue.AccessAndMobilitySubscriptionData.SubscribedUeAmbr = new(models.AmbrRm)
+		// }
```
Comment (andy89923) [inline internal/context/amf_ue.go:688]: @donald1218

File: internal/sbi/processor/notifier/subscription.go [hunk 6]
```diff
@@ -39,15 +39,14 @@ func SendAmfStatusChangeNotify(amfStatus string, guamiList []models.Guami) {
...
 		uri := subscriptionData.AmfStatusUri
 
+		amfStatusNotificationReq := Namf_Communication.AmfStatusChangeNOtifyRequest{
+			AmfStatusChangeNotification: &amfStatusNotification,
+		}
 		logger.ProducerLog.Infof("[AMF] Send Amf Status Change Notify to %s", uri)
-		httpResponse, err := client.AmfStatusChangeCallbackDocumentApiServiceCallbackDocumentApi.
-			AmfStatusChangeNotify(context.Background(), uri, amfStatusNotification)
+		_, err := client.IndividualSubscriptionDocumentApi.
+			AmfStatusChangeNOtify(context.Background(), uri, &amfStatusNotificationReq)
```
Comment (HanHongChen) [inline internal/sbi/processor/notifier/subscription.go:47]: Spell mistake

File: internal/gmm/handler.go [hunk 7]
```diff
@@ -1952,7 +1951,7 @@ func HandleAuthenticationResponse(ue *context.AmfUe, accessType models.AccessTyp
 	}
 
 	switch ue.AuthenticationCtx.AuthType {
-	case models.AuthType__5_G_AKA:
+	case models.AusfUeAuthenticationAuthType__5_G_AKA:
```
[the remark this answers is not in the document]
Comment (tim1207) [reply internal/gmm/handler.go:1954]: This is generated by openapi.

File: internal/sbi/processor/notifier/subscription.go (hunk 6 above)
Comment (tim1207) [reply internal/sbi/processor/notifier/subscription.go:47]: This is generated by openapi.

```
// AmfStatusChangeNOtifyRequest
type AmfStatusChangeNOtifyRequest struct {
	AmfStatusChangeNotification *models.AmfStatusChangeNotification
}
```

File: internal/sbi/consumer/udm_service.go [hunk 8]
```diff
@@ -146,36 +133,28 @@ func (s *nudmService) SDMGetSmfSelectData(ue *amf_context.AmfUe) (problemDetails
...
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		problemDetails = &problem
+		ue.SmfSelectionData = &data.SmfSelectionSubscriptionData
 	} else {
-		err = openapi.ReportError("server no response")
+		err = localErr
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			geterr := apiErr.Model().(Nudm_SubscriberDataManagement.GetSmfSelDataError)
```
[the remark this answers is not in the document]
Comment (tim1207) [reply internal/sbi/consumer/udm_service.go]: fixed

File: internal/gmm/handler.go [hunk 9]
```diff
@@ -1141,8 +1140,9 @@ func handleRequestedNssai(ue *context.AmfUe, anType models.AccessType) error {
 		if needSliceSelection {
 			if ue.NssfUri == "" {
 				for {
+					reqparam := Nnrf_NFDiscovery.SearchNFInstancesRequest{}
```
[the remark this answers is not in the document]
Comment (tim1207) [reply internal/gmm/handler.go]: fixed

File: internal/sbi/consumer/pcf_service.go [hunk 10]
```diff
@@ -104,68 +102,61 @@ func (s *npcfService) AMPolicyControlCreate(
...
-			err = localErr
-			return problemDetails, err
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		problemDetails = &problem
 	} else {
-		err = openapi.ReportError("server no response")
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			reporterr := apiErr.Model().(Npcf_AMPolicy.ReportObservedEventTriggersForIndividualAMPolicyAssociationError)
```
Comment (andy89923) [inline internal/sbi/consumer/pcf_service.go:157]: do the type assertion

File: internal/sbi/consumer/pcf_service.go [hunk 11]
```diff
@@ -104,68 +102,61 @@ func (s *npcfService) AMPolicyControlCreate(
...
-		if httpResp.Status != localErr.Error() {
-			return nil, localErr
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		return &problem, nil
 	} else {
-		return nil, openapi.ReportError("server no response")
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			createrr := apiErr.Model().(Npcf_AMPolicy.CreateIndividualAMPolicyAssociationError)
```
Comment (andy89923) [inline internal/sbi/consumer/pcf_service.go:108]: type assertion 
```suggestion
			creatErr := apiErr.Model().(Npcf_AMPolicy.CreateIndividualAMPolicyAssociationError)
```

File: internal/sbi/consumer/pcf_service.go [hunk 12]
```diff
@@ -176,31 +167,24 @@ func (s *npcfService) AMPolicyControlDelete(ue *amf_context.AmfUe) (problemDetai
...
-			err = localErr
-			return nil, err
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		problemDetails = &problem
 	} else {
-		err = openapi.ReportError("server no response")
+		if apiErr, ok := err.(openapi.GenericOpenAPIError); ok {
+			// API error
+			deleteerr := apiErr.Model().(Npcf_AMPolicy.DeleteIndividualAMPolicyAssociationError)
```
Comment (andy89923) [inline internal/sbi/consumer/pcf_service.go:185]: type assertion 
```suggestion
			deleteErr := apiErr.Model().(Npcf_AMPolicy.DeleteIndividualAMPolicyAssociationError)
```

File: internal/sbi/consumer/smf_service.go [hunk 13]
```diff
@@ -149,60 +149,46 @@ func (s *nsmfService) SelectSmf(
...
+	postSmContextReponse, localErr := client.SMContextsCollectionApi.
+		PostSmContexts(ctx, &postSmContextsRequest)
+	if localErr == nil {
+		smContextRef = postSmContextReponse.Location
 	} else {
-		err1 = openapi.ReportError("server no response")
+		err1 = localErr
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			posterr := apiErr.Model().(Nsmf_PDUSession.PostSmContextsError)
```
Comment (andy89923) [inline internal/sbi/consumer/smf_service.go:181]: type assertion
```suggestion
			postErr := apiErr.Model().(Nsmf_PDUSession.PostSmContextsError)
```

File: internal/sbi/consumer/udm_service.go [hunk 14]
```diff
@@ -146,36 +133,28 @@ func (s *nudmService) SDMGetSmfSelectData(ue *amf_context.AmfUe) (problemDetails
...
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		problemDetails = &problem
+		ue.SmfSelectionData = &data.SmfSelectionSubscriptionData
 	} else {
-		err = openapi.ReportError("server no response")
+		err = localErr
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			getErr := apiErr.Model().(Nudm_SubscriberDataManagement.GetSmfSelDataError)
```
Comment (andy89923) [inline internal/sbi/consumer/udm_service.go:155]: type assertion

File: internal/sbi/consumer/udm_service.go [hunk 15]
```diff
@@ -232,33 +205,28 @@ func (s *nudmService) SDMSubscribe(ue *amf_context.AmfUe) (problemDetails *model
...
-			return problemDetails, err
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		problemDetails = &problem
 	} else {
-		err = openapi.ReportError("server no response")
+		err = localErr
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			subcribeerr := apiErr.Model().(Nudm_SubscriberDataManagement.SubscribeError)
```
Comment (andy89923) [inline internal/sbi/consumer/udm_service.go:227]:
```suggestion
			subcribeErr := apiErr.Model().(Nudm_SubscriberDataManagement.SubscribeError)
```

File: internal/sbi/consumer/udm_service.go [hunk 16]
```diff
@@ -311,15 +273,13 @@ func (s *nudmService) SDMGetSliceSelectionSubscriptionData(
...
-			return problemDetails, err
-		}
-		problem := localErr.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
-		problemDetails = &problem
 	} else {
-		err = openapi.ReportError("server no response")
+		err = localErr
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			geterr := apiErr.Model().(Nudm_SubscriberDataManagement.GetNSSAIError)
```
Comment (andy89923) [inline internal/sbi/consumer/udm_service.go:280]:
```suggestion
			getErr := apiErr.Model().(Nudm_SubscriberDataManagement.GetNSSAIError)
```

File: internal/sbi/consumer/udm_service.go [hunk 17]
```diff
@@ -330,31 +290,25 @@ func (s *nudmService) SDMUnsubscribe(ue *amf_context.AmfUe) (problemDetails *mod
...
+		SubscriptionId: &ue.SdmSubscriptionId,
+	}
+
+	_, localErr := client.SubscriptionDeletionApi.Unsubscribe(ctx, &unsubscribeReq)
+
+	if localErr != nil {
+		err = localErr
+		if apiErr, ok := localErr.(openapi.GenericOpenAPIError); ok {
+			// API error
+			unsubcribeerr := apiErr.Model().(Nudm_SubscriberDataManagement.UnsubscribeError)
```
Comment (andy89923) [inline internal/sbi/consumer/udm_service.go:309]:
```suggestion
			unsubcribeErr := apiErr.Model().(Nudm_SubscriberDataManagement.UnsubscribeError)
```

**Review by ianchen0119 — APPROVED** [review]

**Pull request merged**